datanode/dehistory/initialise/initialise.go

```diff
@@ -0,0 +1,300 @@
+	defer ctxCancelFn()
+
+	var activePeerAddresses []string
+	// Time for connections to be established
+	time.Sleep(5 * time.Second)
+	for retries := 0; retries < 5; retries++ {
+		activePeerAddresses = deHistoryService.GetActivePeerAddresses()
+		if len(activePeerAddresses) == 0 {
+			time.Sleep(5 * time.Second)
+		}
+	}
+
+	if len(activePeerAddresses) == 0 {
+		return fmt.Errorf("failed to find any active peer addresses")
+	}
+
+	mostRecentHistorySegmentFromPeers, err := GetMostRecentHistorySegmentFromPeers(ctx, log.Infof, activePeerAddresses, grpcAPIPort)
+	if err != nil {
+		return fmt.Errorf("failed to get most recent history segment from peers:%w", err)
+	}
+
+	log.Infof("got most recent history segment:%s", mostRecentHistorySegmentFromPeers)
+
+	log.Infof("fetching history using as the first segment:{%s} and minimum block count of %d", mostRecentHistorySegmentFromPeers, cfg.MinimumBlockCount)
+
+	blocksFetched, err := FetchHistoryBlocks(ctx, log.Infof, mostRecentHistorySegmentFromPeers.HistorySegmentId,
+		func(ctx context.Context, historySegmentID string) (FetchResult, error) {
+			segment, err := deHistoryService.FetchHistorySegment(ctx, historySegmentID)
+			if err != nil {
+				return FetchResult{}, err
+			}
+			return FromSegmentIndexEntry(segment), nil
+		}, cfg.MinimumBlockCount)
+
+	log.Infof("fetched %d blocks from decentralised history", blocksFetched)
+
+	log.Infof("loading history into the datanode")
+	from, to, err := deHistoryService.LoadAllAvailableHistoryIntoDatanode(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to load history into the datanode%w", err)
+	}
+	log.Infof("loaded history from height %d to %d into the datanode", from, to)
+
+	return nil
+}
+
+var ErrFailedToGetSegment = errors.New("no history segment found")
+
+func GetMostRecentHistorySegmentFromPeers(ctx context.Context, logInfof func(s string, args ...interface{}), peerAddresses []string,
+	grpcAPIPort int,
+) (*v2.HistorySegment, error) {
+	const maxPeersToContact = 10
+
+	if len(peerAddresses) > maxPeersToContact {
+		peerAddresses = peerAddresses[:maxPeersToContact]
+	}
+
+	ctxWithTimeOut, ctxCancelFn := context.WithTimeout(ctx, 30*time.Second)
+	defer ctxCancelFn()
+	peerToSegment := map[string]*v2.HistorySegment{}
+	for _, peerAddress := range peerAddresses {
+		// We assume here that all/most datanodes will be running their GRPC API on the same port
+		segment, err := GetMostRecentHistorySegmentFromPeer(ctxWithTimeOut, peerAddress, grpcAPIPort)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get most recent history segment:%w", err)
+		}
+		peerToSegment[peerAddress] = segment
+	}
+
+	segmentsList := make([]*v2.HistorySegment, 0, len(peerToSegment))
+
+	segmentsInfo := "Most Recent History Segments:\n\n"
+	for peer, segment := range peerToSegment {
+		segmentsList = append(segmentsList, segment)
+		segmentsInfo += fmt.Sprintf("Peer:%-39s Segment{%s}\n\n", peer, segment)
+	}
+
+	logInfof(segmentsInfo)
```

Comment:
Why do this formatting here? If can we just log info the type directly? adding \n and stuff in a log will be not very readable / practical to use seeing that any other logs are meant to be jsion structured logs

Reply:
good point, sorted that now